src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -317,6 +344,40 @@ private static bool TryReadMessage(ref ReadOnlySequence<byte> buffer, HttpContex
             var messageBuffer = buffer.Slice(HeaderSize, messageLength);
             message = messageBuffer.ToArray();
 
+            if (compressed)
+            {
+                string encoding = context.GetGrpcEncoding();
+                if (encoding == null)
+                {
+                    throw new RpcException(NoMessageEncodingMessageStatus);
+                }
+                if (string.Equals(encoding, GrpcProtocolConstants.IdentityGrpcEncoding, StringComparison.Ordinal))
+                {
+                    throw new RpcException(IdentityMessageEncodingMessageStatus);
+                }
+
+                // Performance improvement would be to decompress without converting to an intermediary byte array
+                if (!GrpcProtocolHelpers.TryDecompressMessage(encoding, context.ServiceOptions.CompressionProviders, message, out var decompressedMessage))
+                {
+                    // https://github.com/grpc/grpc/blob/master/doc/compression.md#test-cases
+                    // A message compressed by a client in a way not supported by its server MUST fail with status UNIMPLEMENTED,
+                    // its associated description indicating the unsupported condition as well as the supported ones. The returned
+                    // grpc-accept-encoding header MUST NOT contain the compression method (encoding) used.
+                    var supportedEncodings = context.ServiceOptions.CompressionProviders.Select(p => p.EncodingName).ToList();
+
+                    if (!context.HttpContext.Response.HasStarted)
+                    {
+                        context.HttpContext.Response.Headers[GrpcProtocolConstants.MessageAcceptEncodingHeader] = string.Join(",", supportedEncodings);
+                    }
+
+                    throw new RpcException(CreateUnknownMessageEncodingMessageStatus(encoding, supportedEncodings));
+                }
+
+                context.ValidateGrpcAcceptEncoding();
```

Comment:
Wait a minute, why do we need to validate the encoding here? As long as the server can decode the request it's fine.

Reply:
https://user-images.githubusercontent.com/2030323/54811447-ef22cb00-4c45-11e9-8b73-23bb296f19be.png

"Is the algorithm present in the incoming grpc-accept-encoding?"

Reply:
Ah, I see. Then I'm really confused as to what the grpc-accept-encoding means then. Let's talk about this during standup.

Reply:
This one just logs a message to make sure the implementation is compliant with a the spec.
perhaps rename the ValidateGrpcAcceptEncoding() to make it clearer what it does?